gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -190,11 +191,70 @@ impl GatewayLightning for ClnRpcService {
             })
     }
 
+    // type PayInvoiceStream = tonic::Streaming<PayInvoiceResponse>;
+    type PayInvoiceStream =
+        Pin<Box<dyn Stream<Item = Result<PayInvoiceResponse, tonic::Status>> + Send + 'static>>;
+
     async fn pay_invoice(
         &self,
-        _request: tonic::Request<PayInvoiceRequest>,
-    ) -> Result<tonic::Response<PayInvoiceResponse>, Status> {
-        unimplemented!()
+        request: tonic::Request<tonic::Streaming<PayInvoiceRequest>>,
+    ) -> Result<tonic::Response<Self::PayInvoiceStream>, tonic::Status> {
+        let mut stream = request.into_inner();
+
+        let mut outcomes: Vec<Result<PayInvoiceResponse, tonic::Status>> = Vec::new();
+
+        // Since the cln_rpc client is not `Sync`, we cannot actually lazily pocess the requests into response stream
+        // TODO: Fix clnrpc lifetime to muntithreading and lazy processing
+        while let Some(PayInvoiceRequest {
+            invoice,
+            max_delay,
+            max_fee_percent,
+        }) = stream.message().await?
```

Comment:
It's unclear to me why invoices to be paid have to come in a stream instead of sigle requests? Does it have to do with the client being in a mutex? If so, why is that? CLN is perfectly able to pay things in parallel.

Reply:
We [defined the grpc spec](https://github.com/fedimint/fedimint/pull/1248) to allow the client to send a stream of invoices to be paid.
While unnecessary at this point, it allows for nice batching of requests by a client, i.e. the client can do `service.pay_invoice([invoice, invoice])` instead of multiple calls to `service.pay_invoice(invoice)`.

This spec only tells the service to assume/expect multiple invoices with every request.
- It's up to the service to optimize how it pays the invoices.
- It might consider parallel processing on every single request (stream from a single client) and across several requests (multiple streams, possibly from multiple clients) should be explored.

Reply:
Curious here, too

Reply:
Two aspects of concern:

- Should the invoices to be paid come in a stream? : Yes - so a client can minimize it's requests (round trips) to the extension service. Sending a stream of invoices to be paid is the spec we want for that. See
`rpc PayInvoice(stream PayInvoiceRequest) returns (stream PayInvoiceResponse)`
- Can / should the extension service pay multiple invoices in parallel? Hopefully yes - Would be cool to parallel process pay invoice requests, but it's all up to the extension impl